File: finops-go/internal/executor/executor.go
package executor

import (
	"fmt"
	"time"

	"github.com/finops-claw-gang/finops-go/internal/domain"
	"github.com/finops-claw-gang/finops-go/internal/policy"
)

// Executor performs deterministic action execution. It takes pre/post
// snapshots and calls the policy safety gate before any action.
type Executor struct {
	tags TagFetcher
}

// NewExecutor creates an Executor backed by the given TagFetcher.
func NewExecutor(tags TagFetcher) *Executor {
	return &Executor{tags: tags}
}

// Snapshot captures the pre- or post-action state for the given action.
// If the action has a target resource, the snapshot includes its tags.
func (e *Executor) Snapshot(action domain.RecommendedAction) (map[string]any, error) {
	if action.TargetResource != "" {
		tags, err := e.tags.ResourceTags(action.TargetResource)
		if err != nil {
			return nil, fmt.Errorf("executor: snapshot tags for %s: %w", action.TargetResource, err)
		}
		return map[string]any{"tags": tags}, nil
	}
	return map[string]any{}, nil
}

// ExecuteActions runs each approved action sequentially, taking pre/post
// snapshots and enforcing the policy safety gate up front.
// Execution stops on the first failure.
func (e *Executor) ExecuteActions(
	approval domain.ApprovalStatus,
	actions []domain.RecommendedAction,
	resourceTagsByARN map[string]map[string]string,
) ([]domain.ExecutionResult, error) {
	if err := policy.EnforceExecutorSafety(approval, actions, resourceTagsByARN); err != nil {
		return nil, err
	}

	results := make([]domain.ExecutionResult, 0, len(actions))
	for _, a := range actions {
		pre, err := e.Snapshot(a)
		if err != nil {
			return nil, fmt.Errorf("executor: pre-snapshot: %w", err)
		}

		// TODO: production impl should deep-copy pre into post and capture real state
		results = append(results, domain.ExecutionResult{
			ActionID:           a.ActionID,
			ExecutedAt:         time.Now().UTC().Format(time.RFC3339),
			Success:            true,
			Details:            fmt.Sprintf("stub executed %s on %s", a.ActionType, a.TargetResource),
			RollbackAvailable:  true,
			PreActionSnapshot:  pre,
			PostActionSnapshot: pre,
		})
	}
	return results, nil
}

File: finops-deterministic/src/finops_desk/executor.py
from __future__ import annotations

from datetime import datetime, timezone

from .models import ApprovalStatus, ExecutionResult, RecommendedAction
from .policy import enforce_executor_safety
from .tools import InfraTools, now_utc_iso


class DeterministicExecutor:
    """no llm. just tools + explicit snapshots."""

    def __init__(self, infra: InfraTools):
        self.infra = infra

    def snapshot(self, action: RecommendedAction) -> dict:
        # in prod: describe_resource, asg config, etc.
        if action.target_resource:
            return {"tags": self.infra.resource_tags(action.target_resource)}
        return {}

    def execute_actions(
        self,
        approval: ApprovalStatus,
        actions: list[RecommendedAction],
        resource_tags_by_arn: dict[str, dict[str, str]],
    ) -> list[ExecutionResult]:
        enforce_executor_safety(approval, actions, resource_tags_by_arn)

        results: list[ExecutionResult] = []

        for a in actions:
            pre = self.snapshot(a)
            # stub: we do not actually call aws
            ok = True
            details = f"stub executed {a.action_type} on {a.target_resource}"
            post = pre

            results.append(
                ExecutionResult(
                    action_id=a.action_id,
                    executed_at=now_utc_iso(),
                    success=ok,
                    details=details,
                    rollback_available=True,
                    pre_action_snapshot=pre,
                    post_action_snapshot=post,
                )
            )

            if not ok:
                break

        return results
